File: binary/Binary.hs
{-# LANGUAGE LambdaCase #-}
{-|
Module      : Binary
Copyright   : (c) Eric Bailey, 2015
License     : MIT

Maintainer  : Eric Bailey
Stability   : experimental
Portability : portable

Converting string binary numbers to their decimal equivalents.
-}

module Binary (
  toDecimal,
  -- * Usage
  -- $usage
  charToBinaryM,
  (<$+>)
  ) where

import           Control.Monad (foldM)
import           Data.Maybe    (fromMaybe)

-- | Given a 'String' representing a binary number, calls 'charToBinaryM' on
-- each character. If any call returns @Nothing@, the fold returns @Nothing@,
-- otherwise folds the list of @Just@ binary digits from left to right,
-- by doubling the accumulator (starting at @0@) and adding the @Just@ value.
-- After the the fold, the resulting @Maybe Int@ is unwrapped,
-- with @0@ as the default.
toDecimal :: String -> Int
toDecimal = fromMaybe 0 . foldM ((. charToBinaryM) . (<$+>) . (2*)) 0

{- $usage
>>> toDecimal "11010"
acc' = (2*) acc <$+> charToBinaryM c -- (   seen,         rest)
-----------------------------------------------------------
   0 = 0                             -- (     "",      "11010")
   1 = (2 *  0) + 1                  -- (    "1",       "1010")
   3 = (2 *  1) + 1                  -- (   "11",        "010")
   6 = (2 *  3) + 0                  -- (  "110",         "10")
  13 = (2 *  6) + 1                  -- ( "1101",          "0")
  26 = (2 * 13) + 0                  -- ("11010",           "")

>>> toDecimal "101"
   0 = 0                             -- (     "",        "101")
   1 = (2 *  0) + 1                  -- (    "1",         "01")
   2 = (2 *  1) + 0                  -- (   "10",          "1")
   5 = (2 *  2) + 1                  -- (  "101",           "")

>>> toDecimal "0xDEADBEEF"
   0 = 0                             -- (     "", "0xDEADBEEF")
   0 = (2 *  0) + 0                  -- (    "0",  "xDEADBEEF")
   ...
   0 = fromMaybe 0 Nothing
-}

-- | Given a 'Char', convert @'0'@ to @Just 0@ and @'1'@ to @Just 1@,
-- otherwise return @Nothing@.
charToBinaryM :: Char -> Maybe Int
charToBinaryM = \case '0' -> Just 0; '1' -> Just 1; _ -> Nothing

-- | Given an 'Int' @x@ and a 'Maybe' @Int y@, if @y@ is a @Just@,
-- return @Just (x + y)@, otherwise @Nothing@.
(<$+>) :: Int -> Maybe Int -> Maybe Int
(<$+>) = (<$>) . (+)

File: binary/binary.clj
(ns binary
  "Converting binary number strings to their decimal equivalents."
  {:author "Eric Bailey"})

;; == NILABLE VERSION ==========================================================

(defn to-decimal
  "Given a string representing a binary number, return its decimal equivalent.

  With 0 as the initial value, perform a left-associative fold over the string,
  using case analysis on each character.  If a character is `\1` or `\0`, add
  its numeric value to double the current result.  Otherwise, when any character
  other than `\0` or `\1` is present, return 0."
  [string]
  (-> (reduce (fn [sum c]
                (when sum
                  (some-> (case c \1 1, \0 0, nil)
                    (+ (* 2 sum)))))
              0 string)
      (or 0)))


;; == MONADIC VERSION ==========================================================

(comment
  ;; http://funcool.github.io/cats/latest/
  (require '[cats.core [>>= foldm]]
           '[cats.monad.maybe :refer [from-maybe just nothing]])

  (defn to-decimal [string]
    (-> (foldm (fn [sum c]
                 (>>= (case c \1 (just 1), \0 (just 0), (nothing))
                      #(just (+ % (* 2 sum)))))
               0 string)
        (from-maybe 0))))
